### ISSUE TITLE
SPDZ2k tests fails on network related issue

### ISSUE THREAD

**Issue by GuutBoy** [aicis/fresco#289]
The SPDZ2k tests occasionally fails with a network related bug. Below is part of the log when this occurs (see https://api.travis-ci.org/v3/job/379235437/log.txt for the entire log):

```
14:24:55.275 [pool-23-thread-2] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P3: bound at 0.0.0.0/0.0.0.0:42253
14:24:55.291 [pool-23-thread-2] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P3: accepted connection from Party(1, localhost:36447)
14:24:55.300 [Thread-6] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P2: Network closed
14:24:55.300 [pool-24-thread-2] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P1: bound at 0.0.0.0/0.0.0.0:36447
14:24:55.291 [pool-24-thread-1] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P1: connected to Party(2, localhost:42253)
14:24:55.301 [pool-24-thread-1] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P1: connected to Party(3, localhost:42253)
14:24:55.301 [pool-25-thread-1] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P2: connected to Party(3, localhost:42253)
14:24:55.302 [pool-23-thread-2] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P3: accepted connection from Party(2, localhost:42253)
14:24:55.302 [pool-25-thread-2] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P2: bound at 0.0.0.0/0.0.0.0:42253
14:24:55.303 [Thread-5] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P1: successfully connected network
14:24:55.305 [Thread-7] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P3: successfully connected network
14:24:55.306 [SCE-6] INFO dk.alexandra.fresco.framework.sce.SecureComputationEngineImpl - Running application: dk.alexandra.fresco.suite.spdz2k.protocols.computations.TestSpdz2kCommitmentComputation$TestCommitment$1$$Lambda$59/1401325233@74459fbc using protocol suite: dk.alexandra.fresco.suite.spdz2k.Spdz2kProtocolSuite128@7a827f44
14:24:55.306 [SCE-5] INFO dk.alexandra.fresco.framework.sce.SecureComputationEngineImpl - Running application: dk.alexandra.fresco.suite.spdz2k.protocols.computations.TestSpdz2kCommitmentComputation$TestCommitment$1$$Lambda$59/1401325233@3bea9d99 using protocol suite: dk.alexandra.fresco.suite.spdz2k.Spdz2kProtocolSuite128@352d6494
14:24:55.314 [Thread-5] ERROR dk.alexandra.fresco.framework.TestThreadRunner - TestThread(dk.alexandra.fresco.suite.spdz2k.resource.Spdz2kResourcePoolImpl@22bc53a3) threw exception during test:
java.lang.RuntimeException: Execution exception when running the application
	at dk.alexandra.fresco.framework.sce.SecureComputationEngineImpl.runApplication(SecureComputationEngineImpl.java:61)
	at dk.alexandra.fresco.framework.sce.SecureComputationEngine.runApplication(SecureComputationEngine.java:37)
	at dk.alexandra.fresco.framework.TestThreadRunner$TestThread.runApplication(TestThreadRunner.java:41)
	at dk.alexandra.fresco.suite.spdz2k.protocols.computations.TestSpdz2kCommitmentComputation$TestCommitment$1.test(TestSpdz2kCommitmentComputation.java:84)
	at dk.alexandra.fresco.framework.TestThreadRunner$TestThread.runTest(TestThreadRunner.java:65)
	at dk.alexandra.fresco.framework.TestThreadRunner$TestThread.run(TestThreadRunner.java:53)
Caused by: java.lang.RuntimeException: P1: Unable to receive from P3
	at dk.alexandra.fresco.framework.util.ExceptionConverter.safe(ExceptionConverter.java:25)
	at dk.alexandra.fresco.framework.network.AsyncNetwork.receive(AsyncNetwork.java:423)
	at dk.alexandra.fresco.framework.sce.evaluator.NetworkBatchDecorator.receive(NetworkBatchDecorator.java:35)
	at dk.alexandra.fresco.framework.network.Network.receiveFromAll(Network.java:50)
	at dk.alexandra.fresco.suite.spdz2k.protocols.natives.InsecureBroadcastProtocol.evaluate(InsecureBroadcastProtocol.java:38)
	at dk.alexandra.fresco.framework.sce.evaluator.BatchedStrategy.evaluateCurrentRound(BatchedStrategy.java:52)
	at dk.alexandra.fresco.framework.sce.evaluator.BatchedStrategy.processBatch(BatchedStrategy.java:40)
	at dk.alexandra.fresco.framework.sce.evaluator.BatchedProtocolEvaluator.eval(BatchedProtocolEvaluator.java:58)
	at dk.alexandra.fresco.framework.sce.SecureComputationEngineImpl.evalApplication(SecureComputationEngineImpl.java:81)
	at dk.alexandra.fresco.framework.sce.SecureComputationEngineImpl.lambda$startApplication$0(SecureComputationEngineImpl.java:69)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.io.IOException: Connection reset by peer
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at dk.alexandra.fresco.framework.network.AsyncNetwork$Receiver.isRunning(AsyncNetwork.java:258)
	at dk.alexandra.fresco.framework.network.AsyncNetwork.lambda$receive$4(AsyncNetwork.java:424)
	at dk.alexandra.fresco.framework.util.ExceptionConverter.safe(ExceptionConverter.java:23)
	... 13 common frames omitted
Caused by: java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:197)
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
	at dk.alexandra.fresco.framework.network.AsyncNetwork$Receiver.call(AsyncNetwork.java:293)
	... 4 common frames omitted
14:24:55.327 [Thread-5] WARN dk.alexandra.fresco.framework.network.AsyncNetwork - P1: A failed sender detected while closing network
14:24:55.327 [Thread-5] WARN dk.alexandra.fresco.framework.network.AsyncNetwork - P1: A failed receiver detected while closing network
14:24:55.328 [Thread-5] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P1: Network closed
14:24:56.087 [main] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P1: Network already closed
14:24:56.088 [main] INFO dk.alexandra.fresco.framework.network.AsyncNetwork - P3: Network closed
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.17 sec <<< FAILURE!
``` 
It appears that P2 and P3 are assigned the same port number. Somehow this seems to trigger the error. The fact that P2 and P3 are assigned the same port number, points to a bug in the `getFreePorts` method on `AbstractSpdz2kTest` apparently can fail to assign unique ports (which is required as the test is running on localhost).

**Comment by GuutBoy** [comment]
I am pretty sure what is happening here is that:

P1 gets its own port, while P2 and P3 are assigned the same port number. 

P1 and P3 starts listening for connections (before P2 starts trying to bind to the same port as P3). 

P1 connects to what it thinks is P2 but really it is P3, P3 accepts connection

P2 connects to P3, P3 accepts the connection

P1 connects to what it thinks is P3. However, since P3 has already accepted two connections it closes the serversocket without accepting the connection.

P1 and P3 now thinks that they have fully connected the network.

P2 now starts listening for connections (this does not fail as P3 has released the port). It will block until a connection is made from P1 (but this will never happen as P1 thinks it already connected to P2). Eventually P2 should timeout waiting for a connection.

Once P1 tries to send/receive from P3 this will fail because the connection was never accepted.

**Comment by GuutBoy** [comment]
Closed in #290
